doc/aaa/radius_authentication.md

```diff
@@ -0,0 +1,634 @@
+3. RADIUS returns Access-Accept with Management-Privilege-Level (MPL)(Goto 4.),
+   or Access-Reject (Authentication Failed).
+4. Access-Accept: The MPL is cached.
+5. A getpwnam_r() call is made to obtain user information.
+6. If the RADIUS user is not found in the local /etc/passwd file, the NSS
+   RADIUS plugin uses the cached MPL information to return user information.
+7. The user has authenticated successfully.
+
+Note:
+1. If an application requests getpwnam_r() before any authentication is
+   done, the RADIUS NSS plugin can be configured to return a least privileged
+   (MPL 1) temporary user info. On subsequent sessions, the user info will
+   be returned per the cached MPL. SSH is an example of such an application.
+   Alternatively, the user can be asked to login to console for the first
+   time to cache the MPL. Subsequent sessions will now use the cached MPL.
+
+Following is the sequence of events during RADIUS authenticated login over
+SSH:
+
+```
+SSH Client              SONiC Device                     RADIUS Server
+                  SSH Server     NSS RADIUS Plugin
+----------------------------------------------------------------------
+
+    |                 |                 |                    |
+    |                 |                 |                    |
+    |  Login(1)       |                 |                    |
+    +---------------->|                 |                    |
+    |                 | getpwnam_r()(2) |                    |
+    |                 +---------------->|                    |
+    |                 |                 |                    |
+    |                 | user info.(3)   |                    |
+    |                 |<----------------+                    |
+    |                 |                 |                    |
+    |                 |                                      |
+    |                 |         Access-Request(4)            |
+    |                 +------------------------------------->|
+    |                 |                                      |
+    |                 |         Access-Accept(5)             |
+    |                 |<-------------------------------------+
+    |                 |                                      |
+    |             (6) |                 |                    |
+    |                 |                 |                    |
+    |  Success(7)     |                 |                    |
+    |<----------------+                 |                    |
+    |                 |                 |                    |
+```
+
+1. A user tries to login through a SSH client.
+2. A getpwnam_r() call is made to obtain user information.
+3. If the RADIUS user is not found in the local /etc/passwd file, the NSS
+   RADIUS plugin uses the cached MPL information to return user information.
+   If no MPL is available, and the many_to_one=a is configured, a least
+   privileged (MPL 1) temporary user info is returned. Otherwise, the
```

Comment:
Does this mean, a Radius user has to login twice to get his  real privilege level/role, in the following two scenarios ?

a) First time login into SONiC switch (implying no entry in /etc/passwd)
b) When user's privilege level has changed at RADIUS server, since his last login into this SONiC switch?

Say user had higher privilege level. He logged in, say twice to ensure he has the right level.  And he logs out.
Say, later his privilege level at server goes down ( which by the way is standard practice)
Now the user logs into the device again.
This time he has entry in /etc/passwd & an entry in MPL cache too. Note: The cache will have the higher privilege level.
Questions:
a) Will he now get a session at higher privilege level ?
b) Is there any MPL cache timeout that would prevent the above ?
c) Even with MPL cache timed out, /etc/passwd would have the user for remote_user_su. How does this impact ?

Trying to understand, all the different behaviors to deal with when switching to Radius.

Reply:
Thanks for the use-cases. These are all for the many_to_one=a (anonymous) use-case.

This case can be addressed by giving each user a distinct user id, and a distinct group id. Thus  the primary group will be distinct on the SONiC device for every RADIUS user. Any changes to the user's privilege level on the RADIUS server will not affect their primary group id. (I'm not in favor of re-privileging a user, since there are the issues of what can be done about existing resources that are already created for  that user. Another approach would be to re-create re-privileged users as a different username. However, having a distinct primary group id does seem to alleviate that situation.)

With these changes, the answers to the questions would be.
a) No, because after authentication the lower privilege level would be cached and used to start the user's session.
b) The cache would not be used after the login. Once the user's session is active the cache has no effect, since the /etc/passwd entry would have been created.
c) The /etc/passwd entry for the RADIUS user would be created with the RADIUS user's name, and the unique primary group created for that user. None of these have any bearing on the privileges assigned to that user. The privileges assigned to the user are based on the supplementary/secondary groups assigned to the user (User Privilege Table "secondary grps" column).

Yes, thanks for considering the different use-cases. I can make these changes to the HLD. Please advise?

Reply:
Changes made. Summary of Changes section describes the major changes.